src/internal/collection/postgres_collection.go

```diff
@@ -351,30 +430,72 @@ func (c *postgresCollection) list(
 		return nil
 	}
 
-	result := &model{}
-	for rows.Next() {
-		if err := rows.StructScan(result); err != nil {
-			return c.mapSQLError(err, "")
+	// To avoid holding the DB connection open for an unknown duration
+	// dictated by the client's callback, we:
+	// (1) buffer the SQL rows
+	// (2) close the connection
+	// (3) apply f(), the client's callback, to results in the buffer
+	// (4) if the buffer was full, re-execute the query, offset by key, and repeat (1)
+	bufferResults := func(rs *sqlx.Rows) ([]*model, int, error) {
```

Comment:
Could we make this function just do everything for getting a batch (get the query string, perform the query, and buffer the rows)? Having the row close being handled by a different function from the one that created it seems like a bad code smell. We could also just inline this in the for loop with the `if err := func() {...}(); err != nil {...}` pattern if the reason this exists is for the defer.